aten/src/ATen/native/native_functions.yaml

```diff
@@ -62,6 +62,12 @@
   manual_cpp_binding: True
   variants: method
 
+# See "[Note] assert_all_true"
+- func: assert_all_true(Tensor self, str msg) -> ()
+  variants: function
+  dispatch:
+    CompositeExplicitAutograd: assert_all_true
+
```

Comment:
Discussed offline with some folks independently (@Chillee, @albanD, @ezyang). General opinion is the following:
- for async asserts, there already is an `_assert_async` function. We can add a string error message to it and call it a day
- for synchronous asserts, which is what is going on in these operations, we can use a combination of at::equal and a new "fail" operator that can be overridden to provide some more context.

There are some constraints that we may choose to worry about in the future:
- Giving `_assert_async` an error message means that the error message gets materialized whether or not the condition is True/False, which is different from existing TORCH_CHECK behavior. This is a tradeoff we'll have to live with because deferring the construction of the error message is difficult.
- A "fail" operator is only really necessary if we want to provide better error messages.